auparse/auparse_test.go

```diff
@@ -123,41 +142,182 @@ func TestExtractKeyValuePairs(t *testing.T) {
 		},
 		{
 			`x='grep "test" file' y=z`,
-			map[string]*field{
+			fieldMap{
 				"x": {`'grep "test" file'`, `grep "test" file`},
 				"y": newField("z"),
 			},
 		},
 		{
 			`x="grep 'test' file" y=z`,
-			map[string]*field{
+			fieldMap{
 				"x": {`"grep 'test' file"`, `grep 'test' file`},
 				"y": newField("z"),
 			},
 		},
 		{
 			`x="grep \"test\" file" y=z`,
-			map[string]*field{
+			fieldMap{
 				"x": {`"grep \"test\" file"`, `grep \"test\" file`},
 				"y": newField("z"),
 			},
 		},
 		{
 			`x='grep \'test\' file' y=z`,
-			map[string]*field{
+			fieldMap{
 				"x": {`'grep \'test\' file'`, `grep \'test\' file`},
 				"y": newField("z"),
 			},
 		},
 	}
 
 	for _, tc := range tests {
-		out := map[string]*field{}
-		extractKeyValuePairs(tc.in, out)
+		out := extractKeyValuePairs(tc.in)
 		assert.Equal(t, tc.out, out, "failed on: %v", tc.in)
 	}
 }
 
+func BenchmarkExtractKeyValuePairs(b *testing.B) {
+	tests := []struct {
+		name string
+		in   string
+		out  fieldMap
+	}{
+		{
+			"bpf message",
+			`arch=c000003e syscall=321 success=yes exit=0 a0=2 a1=c0059cd6e8 comm="pf-host-agent" exe="/optimyze.cloud/prodfiler/pf-host-agent/pf-host-agent" subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 key="bpf" ARCH=x86_64 SYSCALL=bpf`,
+			fieldMap{
+				"arch":    newField("c000003e"),
+				"syscall": newField("321"),
+				"success": {`yes`, `yes`},
+				"exit":    newField("0"),
+				"a0":      newField("2"),
+				"a1":      {"c0059cd6e8", "c0059cd6e8"},
+				"comm":    {`"pf-host-agent"`, `pf-host-agent`},
+				"exe":     {`"/optimyze.cloud/prodfiler/pf-host-agent/pf-host-agent"`, `/optimyze.cloud/prodfiler/pf-host-agent/pf-host-agent`},
+				"subj":    newField("unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023"),
+				"key":     {`"bpf"`, "bpf"},
+			},
+		},
+		{
+			"short message",
+			`x='grep "test" file' y=z`,
+			fieldMap{
+				"x": {`'grep "test" file'`, `grep "test" file`},
+				"y": newField("z"),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		b.Run(tc.name+"_current", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				b.StartTimer()
+				out := extractKeyValuePairs(tc.in)
+				b.StopTimer()
+				if !assert.Equal(b, tc.out, out, "failed on: %v", tc.in) {
```

Comment:
I think it is best to have a separate unit test that checks the output. This way we can keep the loop closely focused on the code being benchmarked and you don't have to use StartTime / StopTimer. Using `StartTime` and `StopTimer` within the loop is not ideal because it breaks the estimated number of iterations (since the  StartTime/StopTimer take a long time relative to extractKeyValuePairs).

When running `go test -run NONE -bench ExtractKeyValuePairs/bpf_message_current -benchtime=10s -benchmem ./auparse/` it takes close to two minutes. With the start/stop and assert removed the same benchmark task ~11s and completes a similar number of iterations (~1.1M).

Reply:
If there are no side-effects the compiler is allowed to elide the code run in the benchmark. This is unlikely to happen here soon, but it is a possibility. A simple conditional will prevent this, something like `if len(out) != len(tc.out) { b.Errorf("unexpected result length: got:%d want:%d", len(out), len(tc.out) }` would do the trick.

Same below.

Reply:
I removed the assertion altogether, because re-reading I found that `TestExtractKeyValuePairs` is already covering that.